project-base/tests/ShopBundle/Functional/Model/Product/ProductVariantCreationTest.php

```diff
@@ -0,0 +1,135 @@
+    {
+        return [
+            [AvailabilityDataFixture::AVAILABILITY_IN_STOCK],
+            [AvailabilityDataFixture::AVAILABILITY_ON_REQUEST],
+            [AvailabilityDataFixture::AVAILABILITY_OUT_OF_STOCK],
+            [AvailabilityDataFixture::AVAILABILITY_PREPARING],
+        ];
+    }
+
+    /**
+     * @dataProvider variantsWithAvailabilitiesCanBeCreatedProvider
+     * @param string $availabilityReference
+     */
+    public function testVariantsWithAvailabilitiesCanBeCreated(string $availabilityReference): void
+    {
+        $productData = $this->productDataFactory->create();
+        $productData->availability = $this->getReference($availabilityReference);
+
+        $mainProduct = $this->productFacade->create($productData);
+        $secondProduct = $this->productFacade->create($productData);
+        $thirdProduct = $this->productFacade->create($productData);
+
+        $mainVariant = $this->productVariantFacade->createVariant($mainProduct, [$secondProduct, $thirdProduct]);
+
+        $this->assertTrue($mainVariant->isMainVariant());
+        $this->assertContainsAllVariants([$mainProduct, $secondProduct, $thirdProduct], $mainVariant);
+    }
+
+    /**
+     * @return array
+     */
+    public function variantsWithStockCanBeCreatedProvider(): array
+    {
+        return [
+            [0, Product::OUT_OF_STOCK_ACTION_EXCLUDE_FROM_SALE, null],
+            [100, Product::OUT_OF_STOCK_ACTION_EXCLUDE_FROM_SALE, null],
+            [0, Product::OUT_OF_STOCK_ACTION_HIDE, null],
+            [100, Product::OUT_OF_STOCK_ACTION_HIDE, null],
+            [0, Product::OUT_OF_STOCK_ACTION_SET_ALTERNATE_AVAILABILITY, AvailabilityDataFixture::AVAILABILITY_IN_STOCK],
+            [100, Product::OUT_OF_STOCK_ACTION_SET_ALTERNATE_AVAILABILITY, AvailabilityDataFixture::AVAILABILITY_IN_STOCK],
+            [0, Product::OUT_OF_STOCK_ACTION_SET_ALTERNATE_AVAILABILITY, AvailabilityDataFixture::AVAILABILITY_ON_REQUEST],
+            [100, Product::OUT_OF_STOCK_ACTION_SET_ALTERNATE_AVAILABILITY, AvailabilityDataFixture::AVAILABILITY_ON_REQUEST],
+            [0, Product::OUT_OF_STOCK_ACTION_SET_ALTERNATE_AVAILABILITY, AvailabilityDataFixture::AVAILABILITY_OUT_OF_STOCK],
+            [100, Product::OUT_OF_STOCK_ACTION_SET_ALTERNATE_AVAILABILITY, AvailabilityDataFixture::AVAILABILITY_OUT_OF_STOCK],
+            [0, Product::OUT_OF_STOCK_ACTION_SET_ALTERNATE_AVAILABILITY, AvailabilityDataFixture::AVAILABILITY_PREPARING],
+            [100, Product::OUT_OF_STOCK_ACTION_SET_ALTERNATE_AVAILABILITY, AvailabilityDataFixture::AVAILABILITY_PREPARING],
+        ];
+    }
+
+    /**
+     * @dataProvider variantsWithStockCanBeCreatedProvider
+     * @param int $quantity
+     * @param string $outOfStockAction
+     * @param string|null $outOfStockAvailabilityReference
+     */
+    public function testVariantsWithStockCanBeCreated(int $quantity, string $outOfStockAction, ?string $outOfStockAvailabilityReference): void
```

Comment:
```suggestion
    public function testVariantsWithStockCanBeCreated(int $quantity, string $outOfStockAction, ?string $outOfStockAvailabilityReference = null): void
```

Reply:
I didn't want to make it optional, just nullable.